warp/handlers/signature_request_p2p.go

```diff
@@ -0,0 +1,159 @@
+)
+
+var _ p2p.Handler = (*SignatureRequestHandlerP2P)(nil)
+
+const (
+	ErrFailedToParse = iota
+	ErrFailedToGetSig
+	ErrFailedToMarshal
+)
+
+// SignatureRequestHandlerP2P serves warp signature requests using the p2p
+// framework from avalanchego. It is a peer.RequestHandler for
+// message.MessageSignatureRequest.
+type SignatureRequestHandlerP2P struct {
+	backend warp.Backend
+	codec   codec.Manager
+	stats   *handlerStats
+}
+
+func NewSignatureRequestHandlerP2P(backend warp.Backend, codec codec.Manager) *SignatureRequestHandlerP2P {
+	return &SignatureRequestHandlerP2P{
+		backend: backend,
+		codec:   codec,
+		stats:   newStats(),
+	}
+}
+
+func (s *SignatureRequestHandlerP2P) AppRequest(
+	ctx context.Context,
+	nodeID ids.NodeID,
+	deadline time.Time,
+	requestBytes []byte,
+) ([]byte, *common.AppError) {
+	// Per ACP-118, the requestBytes are the serialized form of
+	// sdk.SignatureRequest.
+	req := new(sdk.SignatureRequest)
+	if err := proto.Unmarshal(requestBytes, req); err != nil {
+		return nil, &common.AppError{
+			Code:    ErrFailedToParse,
+			Message: "failed to unmarshal request: " + err.Error(),
+		}
+	}
+
+	unsignedMessage, err := avalancheWarp.ParseUnsignedMessage(req.Message)
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    ErrFailedToParse,
+			Message: "failed to parse unsigned message: " + err.Error(),
+		}
+	}
+	parsed, err := payload.Parse(unsignedMessage.Payload)
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    ErrFailedToParse,
+			Message: "failed to parse payload: " + err.Error(),
+		}
+	}
+
+	var sig [bls.SignatureLen]byte
+	switch p := parsed.(type) {
+	case *payload.AddressedCall:
+		// Note we pass the unsigned message ID to GetMessageSignature since
+		// that is what the backend expects.
+		// However, we verify the types and format of the payload to ensure
+		// the message conforms to the ACP-118 spec.
+		sig, err = s.GetMessageSignature(unsignedMessage.ID())
+		if err != nil {
+			s.stats.IncMessageSignatureMiss()
+		} else {
+			s.stats.IncMessageSignatureHit()
+		}
+	case *payload.Hash:
+		sig, err = s.GetBlockSignature(p.Hash)
+		if err != nil {
+			s.stats.IncBlockSignatureMiss()
+		} else {
+			s.stats.IncBlockSignatureHit()
+		}
+	default:
+		return nil, &common.AppError{
+			Code:    ErrFailedToParse,
+			Message: "unknown payload type",
+		}
+	}
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    ErrFailedToGetSig,
+			Message: "failed to get signature: " + err.Error(),
+		}
+	}
+
+	// Per ACP-118, the responseBytes are the serialized form of
+	// sdk.SignatureResponse.
+	resp := &sdk.SignatureResponse{Signature: sig[:]}
+	respBytes, err := proto.Marshal(resp)
+	if err != nil {
+		return nil, &common.AppError{
+			Code:    ErrFailedToMarshal,
+			Message: "failed to marshal response: " + err.Error(),
+		}
+	}
+	return respBytes, nil
+}
+
+func (s *SignatureRequestHandlerP2P) GetMessageSignature(messageID ids.ID) ([bls.SignatureLen]byte, error) {
```

Comment:
This is future looking, but do you want to add a comment here that when the the existing signature request handlers are removed at some point after the E-upgrade, the `backend` can stop storing the `messageID` -> `rawMessage` mapping and instead just store the `messageID` keys? When a message request is received, the check can be if the message ID of the parsed message exists, and assuming it does, the raw message provided in the request can be used to create the signature.